### PR TITLE
fix: prevent repeated indentation of multiline shell strings

### LINKED ISSUE #298: Shell blocks are not formatted idempotently when containing bash -c '...' single-quoted strings
`snakefmt 2.0.0` is not idempotent when formatting a `shell:` block that passes a multi-line single-quoted string to `bash -c`. Each invocation adds 8 spaces of indentation to the content inside the single quotes, causing `snakefmt --check` to always report the file as needing changes.

## Minimal example

```
# test.smk
rule example:
    shell:
        """
        squashfs-mount image.sqsh:/env -- bash -c '
            source /env/bin/activate
            echo hello
            '
        """
```
Running `snakefmt test.smk` twice produces different output each time:
```
$ snakefmt test.smk && snakefmt --check test.smk
[INFO] Writing formatted content to test.smk
[INFO] All done 🎉
[INFO] 1 file(s) would be changed 😬
```
The indentation inside the single-quoted string grows by 8 spaces on each run.

## Workaround
Wrapping the `shell:` block with `# fmt: off` / `# fmt: on` prevents the repeated reformatting.

## Version
`snakefmt, version 2.0.0`

### PR DESCRIPTION
**Pull request by mbhall88**
Enhance `_indent_preserving_heredocs` to scan character-by-character for shell string quote boundaries (`'`, `"`, `\`). Lines starting inside a multiline string literal are now skipped when applying the python-level target indentation. This ensures formatting of multiline strings inside `shell:` blocks is idempotent.

Resolves #298

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved indentation handling for heredocs and multi-line shell strings to avoid incorrect indentation inside single-quoted, double-quoted, and backtick literals.

* **Tests**
  * Added idempotency and edge-case tests for heredocs and multi-line shell literals (escaped quotes/backticks, comments, empty vs non-empty formatting).
  * Tightened a file-write test to only check file modification time.

[Review Change Stack](https://app.coderabbit.ai/change-stack/snakemake/snakefmt/pull/299?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: tests/test_shell_formatter.py [hunk 1]
```diff
@@ -271,3 +271,127 @@ def test_heredoc_with_custom_delimiter_and_newlines(self):
...
+    text = 'echo "line 1 \\" line 2"\n'
+    res = _indent_preserving_heredocs(text, "    ")
+    assert res == '    echo "line 1 \\" line 2"\n'
+
+    # 4. Escaped backslash in normal code
+    text = "echo \\\\\n"
+    res = _indent_preserving_heredocs(text, "    ")
+    assert res == "    echo \\\\\n"
+
+    # 5. Comment boundaries after control characters and inside words
```
Comment (Hocnonsense) [inline tests/test_shell_formatter.py]: Is the test valid?
```log
$ shfmt --version
v3.13.1
$ cat b
echo bar;# control ;
echo bar&# control &
echo bar|# control |
(# control (
)# control )
echo foo#bar
$ shfmt b
b:4:1: `(` must be followed by a statement list
$ cat b
echo bar;
echo bar&
echo bar|
(
)
echo foo
$ shfmt b
b:4:1: `(` must be followed by a statement list
$ cat b
echo bar;
echo bar&
echo bar|
(a
)
echo foo
$ shfmt b
echo bar
echo bar &
echo bar |
        (
                a
        )
echo foo
```

**Review by Hocnonsense — APPROVED** [review]
Sure, I think it works

File: tests/test_shell_formatter.py (hunk 1 above)
Comment (mbhall88) [reply tests/test_shell_formatter.py]: good pickup. thanks
